src/refactorings/change-signature/change-signature.test.ts

```diff
@@ -0,0 +1,404 @@
+        editor.saveUserChoices(userChangePositionOf(0, 1));
+        editor.saveUserChoices(userChangePositionOf(1, 0));
+        await saveOtherFiles(setup, editor);
+
+        await changeSignature(editor);
+
+        const extracted = await editor.codeOf(setup.currentFile.path);
+        expect(extracted).toBe(expected.currentFile.code);
+        await validateOutput(expected, editor);
+      }
+    );
+  });
+
+  it("should show an error message if refactoring can't be made because rest param should be the last", async () => {
+    const code = `
+    function [cursor]aFn(a, ...args) {
+      return args.push(a);
+    }
+
+    aFn(0, 1);
+    `;
+    const editor = new InMemoryEditor(code);
+    await editor.writeIn(new RelativePath("aModule.js"), editor.code);
+    jest.spyOn(editor, "showError");
+    editor.saveUserChoices(userChangePositionOf(0, 1));
+    editor.saveUserChoices(userChangePositionOf(1, 0));
+
+    await changeSignature(editor);
+
+    expect(editor.showError).toBeCalledWith(
+      ErrorReason.CantChangeSignatureException
+    );
+  });
+
+  it("Should order correclty for complex parameters with defaults values", async () => {
+    const { setup, expected } = {
+      setup: {
+        currentFile: `function [cursor]add(a, str, {item}, [value = 1]) {
+            console.log(a, str, item, value, args)
+          }
+
+          add(7, " years", {item: 3}, [1]);`,
+        path: new RelativePath("./aFileWithReferencesInsideSameFile.ts")
+      },
+      expected: {
+        currentFile: `function add([value = 1], {item}, str, a) {
+            console.log(a, str, item, value, args)
+          }
+
+          add([1], {item: 3}, " years", 7);`
+      }
+    };
+
+    const editor = new InMemoryEditor(setup.currentFile);
+    await editor.writeIn(setup.path, editor.code);
+    editor.saveUserChoices(userChangePositionOf(0, 3));
+    editor.saveUserChoices(userChangePositionOf(3, 0));
+    editor.saveUserChoices(userChangePositionOf(1, 2));
+    editor.saveUserChoices(userChangePositionOf(2, 1));
+
+    await changeSignature(editor);
+
+    const extracted = await editor.codeOf(setup.path);
+    expect(extracted).toBe(expected.currentFile);
+  });
+});
+
+function validateOutput(
+  expected: { currentFile: TestSample; otherFiles: TestSample[] },
+  editor: InMemoryEditor
+) {
+  const promises = expected.otherFiles.map(async (file) => {
+    const extracted = await editor.codeOf(file.path);
+    expect(extracted).toBe(file.code);
+  });
+
+  return Promise.all(promises);
+}
+
+function saveOtherFiles(
+  setup: {
+    currentFile: TestSample;
+    otherFiles: TestSample[];
+  },
+  editor: InMemoryEditor
+) {
+  const promises = setup.otherFiles.map(async (file) => {
+    await editor.writeIn(file.path, file.code);
+  });
+
+  return Promise.all(promises);
+}
+
+function userChangePositionOf(startAt: number, endAt: number) {
+  return {
+    label: Date.now.toString(), // we don't care about param name in inMemory
```

Comment:
🍰 In such a situation, I like to make it explicit and put: `label: 'irrelevant'` instead

Reply:
Make sense to me.

Changes done at https://github.com/nicoespeon/abracadabra/pull/725/commits/8eae0dbca6b41b0215134d6c65fc1735a53e6ac2